File: code/packages/java/avl-tree/src/main/java/com/codingadventures/avltree/AVLTree.java
package com.codingadventures.avltree;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public final class AvlTree<T extends Comparable<? super T>> {
    private final AvlNode<T> root;

    public AvlTree() {
        this(null);
    }

    public AvlTree(AvlNode<T> root) {
        this.root = root;
    }

    public static <T extends Comparable<? super T>> AvlTree<T> empty() {
        return new AvlTree<>();
    }

    public static <T extends Comparable<? super T>> AvlTree<T> fromValues(Iterable<T> values) {
        Objects.requireNonNull(values, "values");
        AvlTree<T> tree = empty();
        for (T value : values) {
            tree = tree.insert(value);
        }
        return tree;
    }

    public AvlNode<T> root() {
        return root;
    }

    public AvlTree<T> insert(T value) {
        return new AvlTree<>(insertNode(root, value));
    }

    public AvlTree<T> delete(T value) {
        return new AvlTree<>(deleteNode(root, value));
    }

    public AvlNode<T> search(T value) {
        return searchNode(root, value);
    }

    public boolean contains(T value) {
        return search(value) != null;
    }

    public T minValue() {
        return minValue(root);
    }

    public T maxValue() {
        return maxValue(root);
    }

    public T predecessor(T value) {
        return predecessor(root, value);
    }

    public T successor(T value) {
        return successor(root, value);
    }

    public T kthSmallest(int k) {
        return kthSmallest(root, k);
    }

    public int rank(T value) {
        return rank(root, value);
    }

    public List<T> toSortedArray() {
        return toSortedArray(root);
    }

    public boolean isValidBst() {
        return isValidBst(root);
    }

    public boolean isValidAvl() {
        return isValidAvl(root);
    }

    public int balanceFactor(AvlNode<T> node) {
        return balanceFactorNode(node);
    }

    public int height() {
        return AvlNode.height(root);
    }

    public int size() {
        return AvlNode.size(root);
    }

    @Override
    public String toString() {
        String rootValue = root == null ? "null" : String.valueOf(root.value());
        return "AvlTree(root=" + rootValue + ", size=" + size() + ", height=" + height() + ")";
    }

    public static <T extends Comparable<? super T>> AvlNode<T> searchNode(AvlNode<T> root, T value) {
        AvlNode<T> current = root;
        while (current != null) {
            int comparison = value.compareTo(current.value());
            if (comparison < 0) {
                current = current.left();
            } else if (comparison > 0) {
                current = current.right();
            } else {
                return current;
            }
        }
        return null;
    }

    public static <T extends Comparable<? super T>> AvlNode<T> insertNode(AvlNode<T> root, T value) {
        if (root == null) {
            return new AvlNode<>(value);
        }

        int comparison = value.compareTo(root.value());
        if (comparison < 0) {
            return rebalance(new AvlNode<>(root.value(), insertNode(root.left(), value), root.right()));
        }
        if (comparison > 0) {
            return rebalance(new AvlNode<>(root.value(), root.left(), insertNode(root.right(), value)));
        }
        return root;
    }

    public static <T extends Comparable<? super T>> AvlNode<T> deleteNode(AvlNode<T> root, T value) {
        if (root == null) {
            return null;
        }

        int comparison = value.compareTo(root.value());
        if (comparison < 0) {
            return rebalance(new AvlNode<>(root.value(), deleteNode(root.left(), value), root.right()));
        }
        if (comparison > 0) {
            return rebalance(new AvlNode<>(root.value(), root.left(), deleteNode(root.right(), value)));
        }

        if (root.left() == null) {
            return root.right();
        }
        if (root.right() == null) {
            return root.left();
        }

        Extracted<T> extracted = extractMin(root.right());
        return rebalance(new AvlNode<>(extracted.minimum(), root.left(), extracted.root()));
    }

    public static <T extends Comparable<? super T>> T minValue(AvlNode<T> root) {
        AvlNode<T> current = root;
        while (current != null && current.left() != null) {
            current = current.left();
        }
        return current == null ? null : current.value();
    }

    public static <T extends Comparable<? super T>> T maxValue(AvlNode<T> root) {
        AvlNode<T> current = root;
        while (current != null && current.right() != null) {
            current = current.right();
        }
        return current == null ? null : current.value();
    }

    public static <T extends Comparable<? super T>> T predecessor(AvlNode<T> root, T value) {
        AvlNode<T> current = root;
        T best = null;
        while (current != null) {
            if (value.compareTo(current.value()) <= 0) {
                current = current.left();
            } else {
                best = current.value();
                current = current.right();
            }
        }
        return best;
    }

    public static <T extends Comparable<? super T>> T successor(AvlNode<T> root, T value) {
        AvlNode<T> current = root;
        T best = null;
        while (current != null) {
            if (value.compareTo(current.value()) >= 0) {
                current = current.right();
            } else {
                best = current.value();
                current = current.left();
            }
        }
        return best;
    }

    public static <T extends Comparable<? super T>> T kthSmallest(AvlNode<T> root, int k) {
        if (root == null || k <= 0) {
            return null;
        }

        int leftSize = AvlNode.size(root.left());
        if (k == leftSize + 1) {
            return root.value();
        }
        if (k <= leftSize) {
            return kthSmallest(root.left(), k);
        }
        return kthSmallest(root.right(), k - leftSize - 1);
    }

    public static <T extends Comparable<? super T>> int rank(AvlNode<T> root, T value) {
        if (root == null) {
            return 0;
        }

        int comparison = value.compareTo(root.value());
        if (comparison < 0) {
            return rank(root.left(), value);
        }
        if (comparison > 0) {
            return AvlNode.size(root.left()) + 1 + rank(root.right(), value);
        }
        return AvlNode.size(root.left());
    }

    public static <T extends Comparable<? super T>> List<T> toSortedArray(AvlNode<T> root) {
        List<T> output = new ArrayList<>();
        inOrder(root, output);
        return output;
    }

    public static <T extends Comparable<? super T>> boolean isValidBst(AvlNode<T> root) {
        return validateBst(root, null, null);
    }

    public static <T extends Comparable<? super T>> boolean isValidAvl(AvlNode<T> root) {
        return validateAvl(root, null, null) != null;
    }

    public static <T> int balanceFactorNode(AvlNode<T> node) {
        return node == null ? 0 : AvlNode.height(node.left()) - AvlNode.height(node.right());
    }

    private static <T extends Comparable<? super T>> AvlNode<T> rebalance(AvlNode<T> node) {
        int balance = balanceFactorNode(node);

        if (balance > 1) {
            AvlNode<T> left = node.left();
            if (left != null && balanceFactorNode(left) < 0) {
                left = rotateLeft(left);
            }
            return rotateRight(new AvlNode<>(node.value(), left, node.right()));
        }

        if (balance < -1) {
            AvlNode<T> right = node.right();
            if (right != null && balanceFactorNode(right) > 0) {
                right = rotateRight(right);
            }
            return rotateLeft(new AvlNode<>(node.value(), node.left(), right));
        }

        return node;
    }

    private static <T extends Comparable<? super T>> AvlNode<T> rotateLeft(AvlNode<T> root) {
        AvlNode<T> right = root.right();
        if (right == null) {
            return root;
        }

        AvlNode<T> newLeft = new AvlNode<>(root.value(), root.left(), right.left());
        return new AvlNode<>(right.value(), newLeft, right.right());
    }

    private static <T extends Comparable<? super T>> AvlNode<T> rotateRight(AvlNode<T> root) {
        AvlNode<T> left = root.left();
        if (left == null) {
            return root;
        }

        AvlNode<T> newRight = new AvlNode<>(root.value(), left.right(), root.right());
        return new AvlNode<>(left.value(), left.left(), newRight);
    }

    private static <T extends Comparable<? super T>> Extracted<T> extractMin(AvlNode<T> root) {
        if (root.left() == null) {
            return new Extracted<>(root.right(), root.value());
        }

        Extracted<T> extracted = extractMin(root.left());
        return new Extracted<>(
                rebalance(new AvlNode<>(root.value(), extracted.root(), root.right())),
                extracted.minimum());
    }

    private static <T extends Comparable<? super T>> void inOrder(AvlNode<T> root, List<T> output) {
        if (root == null) {
            return;
        }
        inOrder(root.left(), output);
        output.add(root.value());
        inOrder(root.right(), output);
    }

    private static <T extends Comparable<? super T>> boolean validateBst(AvlNode<T> root, T minimum, T maximum) {
        if (root == null) {
            return true;
        }
        if (minimum != null && root.value().compareTo(minimum) <= 0) {
            return false;
        }
        if (maximum != null && root.value().compareTo(maximum) >= 0) {
            return false;
        }
        return validateBst(root.left(), minimum, root.value())
                && validateBst(root.right(), root.value(), maximum);
    }

    private static <T extends Comparable<? super T>> Validation validateAvl(AvlNode<T> root, T minimum, T maximum) {
        if (root == null) {
            return new Validation(-1, 0);
        }
        if (minimum != null && root.value().compareTo(minimum) <= 0) {
            return null;
        }
        if (maximum != null && root.value().compareTo(maximum) >= 0) {
            return null;
        }

        Validation left = validateAvl(root.left(), minimum, root.value());
        Validation right = validateAvl(root.right(), root.value(), maximum);
        if (left == null || right == null) {
            return null;
        }

        int expectedHeight = 1 + Math.max(left.height(), right.height());
        int expectedSize = 1 + left.size() + right.size();
        if (root.height() != expectedHeight
                || root.size() != expectedSize
                || Math.abs(left.height() - right.height()) > 1) {
            return null;
        }

        return new Validation(expectedHeight, expectedSize);
    }

    private record Extracted<T>(AvlNode<T> root, T minimum) {
    }

    private record Validation(int height, int size) {
    }
}

File: code/packages/kotlin/avl-tree/src/main/kotlin/com/codingadventures/avltree/AVLTree.kt
package com.codingadventures.avltree

data class AvlNode<T : Comparable<T>>(
    val value: T,
    val left: AvlNode<T>? = null,
    val right: AvlNode<T>? = null,
    val height: Int = 0,
    val size: Int = 1,
) {
    companion object {
        fun <T : Comparable<T>> leaf(value: T): AvlNode<T> = AvlNode(value)

        fun <T : Comparable<T>> create(
            value: T,
            left: AvlNode<T>?,
            right: AvlNode<T>?,
        ): AvlNode<T> = AvlNode(value, left, right, 1 + maxOf(height(left), height(right)), 1 + size(left) + size(right))

        fun <T : Comparable<T>> height(node: AvlNode<T>?): Int = node?.height ?: -1

        fun <T : Comparable<T>> size(node: AvlNode<T>?): Int = node?.size ?: 0
    }
}

class AvlTree<T : Comparable<T>>(val root: AvlNode<T>? = null) {
    fun insert(value: T): AvlTree<T> = AvlTree(insertNode(root, value))

    fun delete(value: T): AvlTree<T> = AvlTree(deleteNode(root, value))

    fun search(value: T): AvlNode<T>? = searchNode(root, value)

    fun contains(value: T): Boolean = search(value) != null

    fun minValue(): T? = minValue(root)

    fun maxValue(): T? = maxValue(root)

    fun predecessor(value: T): T? = predecessor(root, value)

    fun successor(value: T): T? = successor(root, value)

    fun kthSmallest(k: Int): T? = kthSmallest(root, k)

    fun rank(value: T): Int = rank(root, value)

    fun toSortedArray(): List<T> = toSortedArray(root)

    fun isValidBst(): Boolean = isValidBst(root)

    fun isValidAvl(): Boolean = isValidAvl(root)

    fun balanceFactor(node: AvlNode<T>?): Int = balanceFactorNode(node)

    fun height(): Int = AvlNode.height(root)

    fun size(): Int = AvlNode.size(root)

    override fun toString(): String {
        val rootValue = root?.value?.toString() ?: "null"
        return "AvlTree(root=$rootValue, size=${size()}, height=${height()})"
    }

    companion object {
        fun <T : Comparable<T>> empty(): AvlTree<T> = AvlTree()

        fun <T : Comparable<T>> fromValues(values: Iterable<T>): AvlTree<T> =
            values.fold(empty()) { tree, value -> tree.insert(value) }

        fun <T : Comparable<T>> searchNode(root: AvlNode<T>?, value: T): AvlNode<T>? {
            var current = root
            while (current != null) {
                current = when {
                    value < current.value -> current.left
                    value > current.value -> current.right
                    else -> return current
                }
            }
            return null
        }

        fun <T : Comparable<T>> insertNode(root: AvlNode<T>?, value: T): AvlNode<T> {
            if (root == null) return AvlNode.leaf(value)

            return when {
                value < root.value -> rebalance(AvlNode.create(root.value, insertNode(root.left, value), root.right))
                value > root.value -> rebalance(AvlNode.create(root.value, root.left, insertNode(root.right, value)))
                else -> root
            }
        }

        fun <T : Comparable<T>> deleteNode(root: AvlNode<T>?, value: T): AvlNode<T>? {
            if (root == null) return null

            return when {
                value < root.value -> rebalance(AvlNode.create(root.value, deleteNode(root.left, value), root.right))
                value > root.value -> rebalance(AvlNode.create(root.value, root.left, deleteNode(root.right, value)))
                root.left == null -> root.right
                root.right == null -> root.left
                else -> {
                    val extracted = extractMin(root.right)
                    rebalance(AvlNode.create(extracted.minimum, root.left, extracted.root))
                }
            }
        }

        fun <T : Comparable<T>> minValue(root: AvlNode<T>?): T? {
            var current = root
            while (current?.left != null) {
                current = current.left
            }
            return current?.value
        }

        fun <T : Comparable<T>> maxValue(root: AvlNode<T>?): T? {
            var current = root
            while (current?.right != null) {
                current = current.right
            }
            return current?.value
        }

        fun <T : Comparable<T>> predecessor(root: AvlNode<T>?, value: T): T? {
            var current = root
            var best: T? = null

            while (current != null) {
                if (value <= current.value) {
                    current = current.left
                } else {
                    best = current.value
                    current = current.right
                }
            }

            return best
        }

        fun <T : Comparable<T>> successor(root: AvlNode<T>?, value: T): T? {
            var current = root
            var best: T? = null

            while (current != null) {
                if (value >= current.value) {
                    current = current.right
                } else {
                    best = current.value
                    current = current.left
                }
            }

            return best
        }

        fun <T : Comparable<T>> kthSmallest(root: AvlNode<T>?, k: Int): T? {
            if (root == null || k <= 0) return null

            val leftSize = AvlNode.size(root.left)
            return when {
                k == leftSize + 1 -> root.value
                k <= leftSize -> kthSmallest(root.left, k)
                else -> kthSmallest(root.right, k - leftSize - 1)
            }
        }

        fun <T : Comparable<T>> rank(root: AvlNode<T>?, value: T): Int {
            if (root == null) return 0

            return when {
                value < root.value -> rank(root.left, value)
                value > root.value -> AvlNode.size(root.left) + 1 + rank(root.right, value)
                else -> AvlNode.size(root.left)
            }
        }

        fun <T : Comparable<T>> toSortedArray(root: AvlNode<T>?): List<T> {
            val output = mutableListOf<T>()
            inOrder(root, output)
            return output
        }

        fun <T : Comparable<T>> isValidBst(root: AvlNode<T>?): Boolean =
            validateBst(root, null, null)

        fun <T : Comparable<T>> isValidAvl(root: AvlNode<T>?): Boolean =
            validateAvl(root, null, null) != null

        fun <T : Comparable<T>> balanceFactorNode(node: AvlNode<T>?): Int =
            if (node == null) 0 else AvlNode.height(node.left) - AvlNode.height(node.right)

        private fun <T : Comparable<T>> rebalance(node: AvlNode<T>): AvlNode<T> {
            val balance = balanceFactorNode(node)

            if (balance > 1) {
                val left = if (node.left != null && balanceFactorNode(node.left) < 0) {
                    rotateLeft(node.left)
                } else {
                    node.left
                }
                return rotateRight(AvlNode.create(node.value, left, node.right))
            }

            if (balance < -1) {
                val right = if (node.right != null && balanceFactorNode(node.right) > 0) {
                    rotateRight(node.right)
                } else {
                    node.right
                }
                return rotateLeft(AvlNode.create(node.value, node.left, right))
            }

            return node
        }

        private fun <T : Comparable<T>> rotateLeft(root: AvlNode<T>): AvlNode<T> {
            val right = root.right ?: return root
            val newLeft = AvlNode.create(root.value, root.left, right.left)
            return AvlNode.create(right.value, newLeft, right.right)
        }

        private fun <T : Comparable<T>> rotateRight(root: AvlNode<T>): AvlNode<T> {
            val left = root.left ?: return root
            val newRight = AvlNode.create(root.value, left.right, root.right)
            return AvlNode.create(left.value, left.left, newRight)
        }

        private fun <T : Comparable<T>> extractMin(root: AvlNode<T>): Extracted<T> {
            if (root.left == null) {
                return Extracted(root.right, root.value)
            }

            val extracted = extractMin(root.left)
            return Extracted(
                rebalance(AvlNode.create(root.value, extracted.root, root.right)),
                extracted.minimum,
            )
        }

        private fun <T : Comparable<T>> inOrder(root: AvlNode<T>?, output: MutableList<T>) {
            if (root == null) return
            inOrder(root.left, output)
            output += root.value
            inOrder(root.right, output)
        }

        private fun <T : Comparable<T>> validateBst(
            root: AvlNode<T>?,
            minimum: T?,
            maximum: T?,
        ): Boolean {
            if (root == null) return true
            if (minimum != null && root.value <= minimum) return false
            if (maximum != null && root.value >= maximum) return false
            return validateBst(root.left, minimum, root.value) && validateBst(root.right, root.value, maximum)
        }

        private fun <T : Comparable<T>> validateAvl(
            root: AvlNode<T>?,
            minimum: T?,
            maximum: T?,
        ): Validation? {
            if (root == null) return Validation(-1, 0)
            if (minimum != null && root.value <= minimum) return null
            if (maximum != null && root.value >= maximum) return null

            val left = validateAvl(root.left, minimum, root.value) ?: return null
            val right = validateAvl(root.right, root.value, maximum) ?: return null
            val expectedHeight = 1 + maxOf(left.height, right.height)
            val expectedSize = 1 + left.size + right.size
            if (root.height != expectedHeight || root.size != expectedSize || kotlin.math.abs(left.height - right.height) > 1) {
                return null
            }
            return Validation(expectedHeight, expectedSize)
        }

        private data class Extracted<T : Comparable<T>>(val root: AvlNode<T>?, val minimum: T)

        private data class Validation(val height: Int, val size: Int)
    }
}
